docs/design/three_datahall.md

```diff
@@ -0,0 +1,204 @@
+  valueFromEnv: "MyFancyZone"
+  valueFromNode: ""
+  topologyKey: "kubernetes.io/hostname"
+```
+
+If multiple fields of the `value`, `valueFromEnv` or `valueFromNode` are set the following order will be used:
+
+1. `value`
+1. `valueFromEnv`
+1. `valueFromNode`
+
+The `topologyKey` will be used for `topologySpreadConstraints` and for `PodAntiAffinity` and only has to be set if `valueFromNode` is empty, otherwise it will default to that value.
+The `FoundationDBClusterFaultDomain` will be deprecated, the `zoneCount` and `zoneIndex` will be read from the new [multi-cluster field](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/docs/design/plugin_multi_fdb_support.md#proposed-design).
+For all `localities` that define a `valueFromEnv` we would add the key to the `--substitute-variable` flag.
+For `valueFromNode` we have to modify the sidecar to allow it to read labels from Kubernetes nodes (see [Related Links](#related-links)) and pass that information to the according new flag.
+This change should provide the most flexibility to the user to define the required/wanted localities.
+We would set `locality_instance_id` and `locality_machineid` to the current defaults but also allow the user to define custom localities.
+After that change we should deprecate `cluster.Spec.DataCenter` and `cluster.Spec.DataHall`, those values never had an affect.
+
+### Deployment model
+
+#### Single regional Kubernetes cluster
+
+The assumption here is that the user has a regional Kubernetes cluster that spreads across at least 3 different AZs.
+To ensure that all Pods are spread evenly across the AZs the operator should use [Pod Topology Spread Constraints](https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints).
+
+```yaml
+  topologySpreadConstraints:
+  - maxSkew: 1
+    # the topologyKey will be read from the locality
+    topologyKey: topology.kubernetes.io/zone 
+    whenUnsatisfiable: DoNotSchedule
+    labelSelector:
+      matchLabels:
+        foundationdb.org/fdb-cluster-name: test-cluster
+        fdb-process-class: <process-class>
+```
+
+The `topologySpreadConstraints` above ensures that all Pods are evenly spread across the different `topology.kubernetes.io/zone` and must be set by the operator.
+If a regional cluster contains more than 3 AZ and a user only want to use 3 specific AZs the user has to define an additional `NodeAffinity`:
+
+```yaml
+  affinity:
+    nodeAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+        nodeSelectorTerms:
+        - matchExpressions:
+          - key:  topology.kubernetes.io/zone 
+            operator: In
+            values:
+            - az1
+            - az2
+            - az3
+```
+
+The `NodeAffinity` must be set by the user and the operator doesn't take any action to automatically set the value.
+The user would only require to create one `FoundationDBCluster` and set the `redundancy_mode` to `three_data_hall` or `three_datacenter`.
```

Comment:
Is the user required to define locality rules as part of this configuration, or are there reasonable defaults we can apply?

Reply:
Conclusion from discussion: emit an event if they don't have valid locality fields for their redundancy mode.